File: tests/test_admin.js
var assert = require('chai').assert,
    username = process.env.ESC_ADMIN_USERNAME,
    password = process.env.ESC_ADMIN_PASSWORD,
    /* Change the baseURL to your application URL */
    baseURL = "https://escgroup-1.herokuapp.com",
    driver;



const { Builder, By, Key, until } = require('selenium-webdriver');
const { expect } = require('chai');
/**
 * Basic user pages. Checks `title` of routes
 */
describe('Title pages for Admin', function() {
    this.timeout(15000) // because of passport possibly lagging, we need to have a higher timeout rate
    beforeEach(async function (){
        driver = new Builder().forBrowser('chrome').build();
        await driver.getSession().then(sessionid => {
            driver.sessionID = sessionid.id_;
        });
        await driver.get(baseURL);
        await driver.findElement(By.id('btn_login')).click();     // enter login page
        await driver.findElement(By.name('email')).sendKeys(username);         // enter default user email into `email` box
        await driver.findElement(By.name('password')).sendKeys(password, Key.ENTER);         // enter default user password into `password` box
    });

    afterEach(function (done) {
        // driver.executeScript("sauce:job-result=" + (true ? "passed" : "failed"));    // idk what this is used for...
        driver.quit();
        done();
    });

    /**
     * test admin that is logged in
     */
    describe('admin logged in', function(){
        it('should login', async function(){
            try{
                await driver.findElement(By.id('user_logged')).isDisplayed().then(bool => {
                    expect(bool).to.be.true;
                })
            } catch (err) {
                console.log(err);
            }
        });
        describe('/tickets/:user_id/1', function(){
            it('should-open-users-ticket-inprogress-page', async function() {
                await driver.findElement(By.id('btn_tickets')).click();     // enter queue page for the first time
                await driver.findElement(By.id('tickets1')).click();        // click on `in-progress` Button
                try{
                    await driver.getTitle().then(title => {
                        expect(title).to.be.equal("Tickets - In Progress");
                    });
                } catch(err){
                    console.log(err);
                }
            });
        });
        describe('/tickets/:user_id/2', function(){
            it('should-open-users-ticket-solved-page', async function() {
                await driver.findElement(By.id('btn_tickets')).click();     // enter queue page for the first time
                await driver.findElement(By.id('tickets2')).click();        // click on `solved` Button
                try{
                    await driver.getTitle().then(title => {
                        expect(title).to.be.equal("Tickets - Solved");
                    });
                } catch(err){
                    console.log(err);
                }
            });
        });
        describe('/tickets/:ticket_id/0', function(){
            it('should-open-users-ticket-queue-page', async function() {
                await driver.findElement(By.id('btn_tickets')).click();     // enter queue page for the first time
                await driver.findElement(By.id('tickets0')).click();        // click on `solved` Button
                try{
                    await driver.getTitle().then(title => {
                        expect(title).to.be.equal("Tickets - Queued");
                    });
                } catch(err){
                    console.log(err);
                }
            });
        });
        describe('/ (after logout)', function(){
            it('should-be-able-to-log-out', async function() {
                await driver.findElement(By.id('btn_logout')).click();     // enter queue page for the first time
                try{
                    await driver.getTitle().then(title => {
                        expect(title).to.be.equal("Accenture's ACNAPI Portal");
                    });
                } catch(err){
                    console.log(err);
                }
            });
        });
    });
});

File: tests/java/TestAdmin.java
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class TestAdmin {
    private final String USERNAME = System.getenv("ESC_ADMIN_USERNAME");
    private final String PASSWORD = System.getenv("ESC_USER_PASSWORD");
    private final String PHONENUMBER = System.getenv("ESC_PHONENUMBER");
    private final String baseURL = "https://escgroup-1.herokuapp.com";


    private WebDriver driver;
    private WebElement navbar;
    private WebDriverWait wait;

    @BeforeEach
    public void setUp() {
        System.setProperty(DriverProperty.getWebdriver(), DriverProperty.getPathtoDriver());
        driver = new ChromeDriver();
        driver.get(baseURL);
        driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
        wait = new WebDriverWait(driver, 10);

        /* login to user account first */
        navbar = driver.findElement(By.className("navbar"));
        navbar.findElement(By.id("btn_login")).click();
        driver.findElement(By.name("email")).sendKeys(USERNAME);
        driver.findElement(By.name("password")).sendKeys(PASSWORD, Keys.ENTER);
    }

    @AfterEach
    public void tearDown(){
        driver.quit();
        navbar = null;
    }

    /************** Test Cases **************/

    /**
     * Test case: 1 -- check that the ticket pages are working
     */
    @Test
    public void testTicketClickables(){
        navbar = driver.findElement(By.className("navbar"));
        navbar.findElement(By.id("btn_tickets")).click();
        // get all the links
        List<WebElement> tickets = driver.findElements(By.tagName("tr"));
        Collections.shuffle(tickets);
        for (int i = 0; i < tickets.size(); i = i + 1) {
            if (tickets.get(i).getAttribute("href") == null)
                continue;
            boolean staleElementLoaded = true;
            while (staleElementLoaded) {
                try {
                    driver.navigate().to(tickets.get(i).getAttribute("href"));
                    Thread.sleep(1000);
                    driver.navigate().back();
                    tickets = driver.findElements(By.tagName("tr"));
                    Collections.shuffle(tickets);
                    System.out.println("*** Navigated to" + " " + tickets.get(i).getAttribute("href"));
                    staleElementLoaded = false;
                } catch (StaleElementReferenceException e) {
                    staleElementLoaded = true;
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
        driver.findElement(By.id("tickets1")).click();
        // get all the links
        tickets = driver.findElements(By.tagName("tr"));
        Collections.shuffle(tickets);
        for (int i = 0; i < tickets.size(); i = i + 1) {
            if (tickets.get(i).getAttribute("href") == null)
                continue;
            boolean staleElementLoaded = true;
            while (staleElementLoaded) {
                try {
                    driver.navigate().to(tickets.get(i).getAttribute("href"));
                    Thread.sleep(1000);
                    driver.navigate().back();
                    tickets = driver.findElements(By.tagName("tr"));
                    Collections.shuffle(tickets);
                    System.out.println("*** Navigated to" + " " + tickets.get(i).getAttribute("href"));
                    staleElementLoaded = false;
                } catch (StaleElementReferenceException e) {
                    staleElementLoaded = true;
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
        driver.findElement(By.id("tickets2")).click();
        // get all the links
        tickets = driver.findElements(By.tagName("tr"));
        Collections.shuffle(tickets);
        for (int i = 0; i < tickets.size(); i = i + 1) {
            if (tickets.get(i).getAttribute("href") == null)
                continue;
            boolean staleElementLoaded = true;
            while (staleElementLoaded) {
                try {
                    driver.navigate().to(tickets.get(i).getAttribute("href"));
                    Thread.sleep(1000);
                    driver.navigate().back();
                    tickets = driver.findElements(By.tagName("tr"));
                    Collections.shuffle(tickets);
                    System.out.println("*** Navigated to" + " " + tickets.get(i).getAttribute("href"));
                    staleElementLoaded = false;
                } catch (StaleElementReferenceException e) {
                    staleElementLoaded = true;
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    /**
     * Test case: 2 -- test ticket form (no bugs found)
     */
    @Test
    public void testTicketForm(){
        navbar = driver.findElement(By.className("navbar"));
        navbar.findElement(By.id("btn_ticket-form")).click();
        assertEquals("Creation process: Ticket Basics", driver.getTitle());

        driver.navigate().to(baseURL+"/ticket_form/details");
        assertEquals("ACNAPI Ticket Form", driver.getTitle());

        WebElement tinymce_frame = driver.findElement(By.tagName("iframe"));

        Select topics = new Select(driver.findElement(By.name("topic")));
        List<WebElement> topicOptions = topics.getOptions();
        topics.selectByIndex(new Random().nextInt(topicOptions.size()));
        String title = GA.GA.createRubbishTxt();
        String content = GA.GA.createRubbishTxt();
        driver.findElement(By.name("title")).sendKeys(title);
        driver.switchTo().frame(tinymce_frame);
        driver.findElement(By.tagName("body")).sendKeys(content);
        driver.switchTo().defaultContent();

        driver.findElement(By.className("ghost-round")).click();

        wait.until(ExpectedConditions.titleIs("Accenture's ACNAPI Portal"));
        assertEquals("Accenture's ACNAPI Portal", driver.getTitle());
        navbar = driver.findElement(By.className("navbar"));
        navbar.findElement(By.id("btn_tickets")).click();

        wait.until(ExpectedConditions.textToBePresentInElement(driver.findElement(By.tagName("table")),title));
        System.out.println(driver.getPageSource().contains(title));
        assertTrue(driver.getPageSource().contains(title));
    }
}
